toolkit/tools/scheduler/schedulerutils/printresults.go

```diff
@@ -223,20 +231,38 @@ func buildResultsSetToNodesSet(statesSet map[string]*BuildResult) (result map[st
 	return
 }
 
-func getSRPMsState(pkgGraph *pkggraph.PkgGraph, buildState *GraphBuildState) (failedSRPMs map[string]*BuildResult, prebuiltSRPMs, prebuiltDeltaSRPMs, builtSRPMs map[string]bool, blockedSRPMs map[string]*pkggraph.PkgNode) {
+func getSRPMsState(fullGraph, optimizedGraph *pkggraph.PkgGraph, buildState *GraphBuildState) (failedSRPMs map[string]*BuildResult, prebuiltSRPMs, prebuiltDeltaSRPMs, builtSRPMs map[string]bool, blockedSRPMs map[string]*pkggraph.PkgNode, wastedSRPMs map[string]bool) {
 	failedSRPMs = make(map[string]*BuildResult)
 	prebuiltSRPMs = make(map[string]bool)
 	prebuiltDeltaSRPMs = make(map[string]bool)
 	builtSRPMs = make(map[string]bool)
 	blockedSRPMs = make(map[string]*pkggraph.PkgNode)
+	wastedSRPMs = make(map[string]bool)
 
 	for _, failure := range buildState.BuildFailures() {
 		if failure.Node.Type == pkggraph.TypeLocalBuild {
 			failedSRPMs[failure.Node.SrpmPath] = failure
 		}
 	}
 
-	for _, node := range pkgGraph.AllBuildNodes() {
+	optimizedSrpmPaths := make(map[string]bool)
+	for _, node := range optimizedGraph.AllBuildNodes() {
+		logger.Log.Warnf("%v:%v", buildState.nodeToState[node], node)
+		if buildState.IsNodeProcessed(node) {
+			optimizedSrpmPaths[node.SrpmPath] = true
+		}
+	}
+
+	for _, node := range fullGraph.AllBuildNodes() {
+		// check if path was used in the optimized graph
+		nodeInOptimizedGraph := optimizedGraph.HasNode(node)
+		srpmUsed := optimizedSrpmPaths[node.SrpmPath]
+		if !nodeInOptimizedGraph {
+			if !srpmUsed && buildState.IsNodeAvailable(node) {
```

Comment:
Can a node be cached and available? In that case, we didn't really waste any time building it...

Is that why we check srpmUsed? Can we not instead check IsNodeCached(node) so that there is not some undocumented invariant (e.g. cached nodes did not mark their SrpmPath used)

Reply:
I'd consider the wasted build output a debugging tool, I'm probably not going to include it in the final PR. Left it in for now since its useful though to get a rough idea of improvements. My local builds I do for data will have zero cached nodes used.